### PR TITLE
feat: Table widget - new column type select

### LINKED ISSUE #6766: [Feature]: Support Select as a column type in table widget
## Summary

As a dev, I should be able to add the following as column types
- Select

# Success Metrics

| Goal  | Metric |
| ------------- | ------------- |
| Operate on data using input elements | User should be able to perform add/edit action from row by using select elements |

# Requirements

`B:`
| Requirement   | User Story | Importance | Notes |
| ------------- | ------------- | ------------- | ------------- |
| Column type: `select` | Sets the column type to select | High  | - |
| Computed value | The options to be displayed in the select field; type<Array> |  High | - |
| Default value | The option to be selected by default; type<String> | High | - |
| Inherit attr by data type from the the parent widget | All properties from the parent widget should be inherited under each data type here | high | - |
| Disabled | Should the input remain disabled | high | - |
| onOptionChange | trigger event on change of an option | High | - |

# Developer Handoff Document in Figma:
None required

### PR DESCRIPTION
**Pull request by vibhandikyash**
## Description

> Enabled a new Column type select where the user can setup a select widget in a column.

Fixes #6766 

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

> Tested locally.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: FEATURE/6766-new-column-type-select 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.01 **(-0.04)** | 36.56 **(0)** | 34.49 **(0)** | 55.53 **(-0.03)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.94 **(0.23)** | 41.67 **(0.84)** | 36.21 **(0)** | 56.99 **(0.25)**
 :red_circle: | app/client/src/widgets/TableWidget/component/TableStyledWrappers.tsx | 87.06 **(0.16)** | 53.87 **(-1.02)** | 80 **(0.37)** | 87.06 **(0.16)**
 :green_circle: | app/client/src/widgets/TableWidget/component/TableUtilities.tsx | 38.07 **(-0.03)** | 25.22 **(1.33)** | 15 **(-1.22)** | 36.42 **(0.06)**
 :red_circle: | app/client/src/widgets/TableWidget/widget/derived.js | 64.81 **(-0.28)** | 60.92 **(-0.62)** | 51.16 **(0)** | 62.81 **(-0.27)**
 :green_circle: | app/client/src/widgets/TableWidget/widget/getTableColumns.tsx | 56 **(25.57)** | 23.08 **(23.08)** | 33.33 **(33.33)** | 50 **(30)**
 :red_circle: | app/client/src/widgets/TableWidget/widget/index.tsx | 28.54 **(-1.1)** | 9.42 **(-0.06)** | 29.09 **(-0.32)** | 28.11 **(-1.18)**
 :red_circle: | app/client/src/widgets/TableWidget/widget/propertyConfig.ts | 87.27 **(-8.65)** | 100 **(0)** | 81.08 **(-12.47)** | 87.27 **(-8.65)**
 :green_circle: | app/client/src/widgets/TableWidget/widget/propertyUtils.ts | 28.99 **(0.52)** | 10.78 **(0)** | 26.67 **(0)** | 27.91 **(0.57)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
@vibhandikyash semantics are not right!

**Review by marks0351 — CHANGES_REQUESTED** [review]
@vibhandikyash pls add test cases and check if docs need update.
cc: @somangshu

**Review by marks0351 — APPROVED** [review]

**Review by marks0351 — APPROVED** [review]

**Comment by riodeuno** [comment]
UX Issues
- When selecting an option from the select cell, the row gets selected, and triggers the onRowSelected. Let's not propagate these events.
- `currentRow` is not accessible in the property controls of the select type columns
- Server side filtering doesn't need to be in the list of property controls for now. Please remove all references to this control and any functionality which has been added related to this.
- Even if `currentRow` is available to bind, it is not defined when evaluating triggers.
- In all select components, the first option is always shown as highlighted in the list, while there is no selection @Tooluloope 
Screenshot 2021-09-29 at 10 22 57 AM

File: app/client/src/widgets/TableWidget/component/Constants.ts [hunk 1]
```diff
@@ -175,6 +175,7 @@ export interface ColumnProperties {
   boxShadowColor?: string;
   iconButtonStyle?: ButtonStyleType;
   isCellVisible?: boolean;
+  defaultOptionValue?: string;
```
Comment (riodeuno) [inline app/client/src/widgets/TableWidget/component/Constants.ts:178]: Please verify if `options` needs to be here.

File: app/client/src/widgets/TableWidget/component/TableUtilities.tsx [hunk 2]
```diff
@@ -703,3 +704,65 @@ export const renderDropdown = (props: {
     </div>
   );
 };
+
+const StyledDropDownComponent = styled.div`
+  width: 100%;
+`;
+
+export function SelectCell(props: {
```
Comment (riodeuno) [inline app/client/src/widgets/TableWidget/component/TableUtilities.tsx:712]: Please add a detailed comment at the top of this function describing what this does.

File: app/client/src/widgets/TableWidget/component/TableUtilities.tsx [hunk 3]
```diff
@@ -703,3 +704,65 @@ export const renderDropdown = (props: {
...
+      isHidden={props.isHidden}
+    >
+      <StyledDropDownComponent>
+        <DropDownComponent
+          disabled={Boolean(props.isDisabled)}
+          height={0}
+          isFilterable={false}
+          isLoading={false}
+          onFilterChange={noop}
+          onOptionSelected={(optionSelected) => {
```
Comment (riodeuno) [inline app/client/src/widgets/TableWidget/component/TableUtilities.tsx:750]: Let's use `useCallback` here.

File: app/client/src/widgets/TableWidget/widget/derived.js [hunk 4]
```diff
@@ -97,15 +97,25 @@ export default {
...
+      return props.tableData.map((entry, index) => {
         const sanitizedData = {};
 
         for (const [key, value] of Object.entries(entry)) {
           const sanitizedKey = key
             .split(separatorRegex)
             .join("_")
             .slice(0, 200);
-          sanitizedData[sanitizedKey] = value;
+          if (_.has(props.editedColumnData, `${sanitizedKey}.${index}`)) {
```
Comment (riodeuno) [inline app/client/src/widgets/TableWidget/widget/derived.js:108]: Let's add comments to describe what is being computed here.
Let's also add jest tests for those computations in the `derived.test.js` file.

**Review by riodeuno — CHANGES_REQUESTED** [review]

**Comment by Tooluloope** [comment]
> UX Issues
> - In all select components, the first option is always shown as highlighted in the list, while there is no selection @Tooluloope 
> 

@riodeuno The reason for that is for keyboard accessibility. So when the dropdown is active, the first item is auto focused. There is a difference between the focused state and the select state of the options.

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash Tested this PR and Please find the status below,

1.The order in the Select Column Property pane should match the Select widget order 

* Options
*  Default Options
*  Placeholder
* Required Toggle 
* Visible Toggle 
* Disabled Toggle 
* Filterable Toggle 
* OnOption Change Action

2.  Required Toggle is not available 

3. Filterable Toggle is not available 

4. Once when the disabled toggle is enabled and disabled - Then the select column is in disabled mode 

https://loom.com/share/3d22bc4621064815a20da68fed699d8d

5.By Default - Options, Default Options and PlaceHolder are empty - Due to which empty checkbox is displayed on the select column - **Default values can be displayed** 

https://www.loom.com/share/c581beefec334a05954e9393bf7e1fae

**Comment by somangshu** [comment]
Feedback @vibhandikyash:
- Default value needs to change to computed value and should be re-ordered below column type
- Just like in the button column type we have the button properties separated, we will do the same for this column type
- Point 4 from @YogeshJayaseelan feedback is true
- Point 5 from @YogeshJayaseelan feedback should be looked into, Set the defaults as per other column types
- Re-order the disable property to the top of Select properties 
- Documentation changes required here
- Do we have the basic cypress and jest tests for this?


Revert on the above feedback:
- @YogeshJayaseelan since this is in a table and the value is bound the query I dont see the case of using `required` (please explain if required), However in extension to the inline edit in table widget we could introduce some validation here. 
- @YogeshJayaseelan lets raise filterable as a separate requirement later, I would like to see some request around this before taking it up

**Comment by Nikhil-Nandagopal** [comment]
- Triggered row didn't work for me
- The select option was very slow to update
- There were options missing like server side filtering

**Comment by riodeuno** [comment]
@vibhandikyash Let's implement the `triggeredRow` feature here. We populate the `triggeredRow` when we a button cell is clicked. The same should happen for the onOptionChange of the select column type.

We should also include the description of `triggeredRow` in the documentation

**Comment by riodeuno** [comment]
Here are two issues I found. I'm not sure if this is related to the changes in this PR. Please verify.
-  [ ] Math.random(), triggers updates repeatedly
-  [ ] `currentRow` works in an unexpected manner in the onOptionChange binding.
https://user-images.githubusercontent.com/103687/140047799-9913a714-c024-496e-b738-b6605eb4ba1d.mov

https://user-images.githubusercontent.com/103687/140047816-2c7b9d67-70ed-4c67-b30f-36756e7f3902.mov

**Review by riodeuno — CHANGES_REQUESTED** [review]
Please refer to the comments for the requested changes.

**Comment by riodeuno** [comment]
@vibhandikyash  Nice work! This is close to completion. Quite a few problems have been resolved, however, one issue remains. Please see attached screen recording.
I'll get back to you on the debugger issue.

https://user-images.githubusercontent.com/103687/143436046-c2249924-f4b9-4ee1-a6c6-51f39a6d2a71.mov

**Comment by riodeuno** [comment]
@vibhandikyash I'm not able to check the deploy preview as there is a typescript error while led to a failed build. Please fix this and pull the latest release to see if there are any updates to the binding issue.

**Review by riodeuno — DISMISSED** [review]
Looks good to me. We can go ahead with the QA. I'm looking into the debugger issue w.r.t `derivedColumns`

**Review by riodeuno — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash Tested this feature and please find the status below,

Issue 1 - Property fields order will be based on table widget
Issue 2 - Not in scope
Issue 3 - Not in scope
Issue 4 - Fixed,
Issue 5 - **Not Fixed - Default value is not displayed - The select column is empty by default**

Issue 6 - Math.random(), triggers updates repeatedly - This is still happening and what is expected behaviour here @riodeuno - https://loom.com/share/8a9c157790714f1a879c9edc1163bb7b

New issue :

Issue 7 - Text alignment and vertical alignment is not working for select column type - Should alignment work for select column too? @somangshu @vibhandikyash 
https://www.loom.com/share/b1acdfd665e3442db8a9a62cf977fe45

**Comment by somangshu** [comment]
> Issue 7 - Text alignment and vertical alignment is not working for select column type - Should alignment work for select column too? @somangshu @vibhandikyash

@YogeshJayaseelan yes that should be the case

**Comment by vibhandikyash** [comment]
ok-to-test sha=2e5f161

**Comment by vibhandikyash** [comment]
Closing this PR. Feature will be reopened with TableV2.

**Pull request closed without merging**